### PR TITLE
Store collapsed-folder settings

### PR DESCRIPTION
**Pull request by brueckner**
_Hi! This is my first Open Source PR and I am excited! If you have any suggestions, they are warmly welcomed!_

---

## Problem
The UI doesn't store the last (un-)collapsed state of the folders for an account. If the user is making heavy use of email folders, they'll most likely want to have them uncollapsed at all times. At the moment they have to click on "Show all folders" on every reload/revisit.

## Suggested Solution
Store the last chosen state in the database.

## Approach
First I was thinking about a simple flag in `oc_mail_accounts` and while that would certainly work, it would "pollute" the table with UI settings, which gets worse the more settings are added down the line. Instead I stored the settings in `oc_preferences`, which is already used by the mail app to store the `collect-data` flag for example.

I though it might be useful to be able to store more than just the state of collapsible folders down the line, so tried to make it extendable by storing a JSON object like that:

| userid   | appid | configkey                   | configvalue |
|----------|-------|------------------------|--------------------------|
| username | mail  | account-settings | `{"1":{"accountId":1,"collapsed":false,"some-other-setting":true},"2":{"accountId":2,"collapsed":true}}`|

And that's basically it. The `account-settings` are now stored when using the toggle button. Upon reload, the [`PageController.php`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-70e5309fcd3311d771b3db9c93490ea270fd5894769093765ec37edb68e5dd9b) loads the settings and attaches them to [`templates/index.php`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-5258161aaab90de3916c094462a8b5035d08c02eff481989342aeca0a45c4be5) just like `collect-data` etc. The settings are then picked up in [`main.js`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79) where they're merged with the actual account object. 

---

### Additional notes

- I wanted to pass an empty array as a default fallback to `getPreference()` in [`PageController.php`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-70e5309fcd3311d771b3db9c93490ea270fd5894769093765ec37edb68e5dd9b) like this: `$this->preferences->getPreference('account-settings', []);` – however... the return value would be `null` instead of the empty array. Is that intentional? Seems like a bug to me.
- I wasn't sure about the use of ternary operators here and there in terms of coding standards for this project. Please let me know if you want me to refactor these. 
- I wasn't able to set up the integration tests properly yet, so I can't tell if they would pass or not. Any help is much appreciated!

Fixes https://github.com/nextcloud/mail/issues/1025

### CODE REVIEW

**Comment by ChristophWurst** [comment]
> I wanted to pass an empty array as a default fallback to `getPreference()` in [`PageController.php`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-70e5309fcd3311d771b3db9c93490ea270fd5894769093765ec37edb68e5dd9b) like this: `$this->preferences->getPreference('account-settings', []);` – however... the return value would be `null` instead of the empty array. Is that intentional? Seems like a bug to me.

Indeed. This should work.

> I wasn't able to set up the integration tests properly yet, so I can't tell if they would pass or not. Any help is much appreciated!

You can use the `make start-docker` to pull the test image and run it locally. Then you can use `composer run test:integration:dev` to invoke phpunit. This should work.

File: src/components/Navigation.vue [hunk 1]
```diff
@@ -43,7 +43,7 @@
 									!group.account.collapsed ||
 									!isCollapsed(group.account, item)
 							"
-							:key="item.databaseId"
+							:key="'mailbox-' + item.databaseId"
```
Comment (ChristophWurst) [inline src/components/Navigation.vue:46]: looks like an unrelated change, no? and the databaseId should be unique enough. Did you just add this for clarity or to fix a bug?

File: src/components/NavigationAccountExpandCollapse.vue [hunk 2]
```diff
@@ -53,6 +54,16 @@ export default {
 	methods: {
 		toggleCollapse() {
 			this.$store.commit('toggleAccountCollapsed', this.account.id)
+			this.$store
+				.dispatch('updateAccountSetting', {
+					accountId: this.account.id,
+					key: 'collapsed',
+					value: this.account.collapsed
```
Comment (ChristophWurst) [inline src/components/NavigationAccountExpandCollapse.vue:61]: tip: we use ESLint for the js formatting. You can `npm run lint:fix` to have everything in the desired format automatically :)

**Comment by brueckner** [comment]
> You can use the `make start-docker` to pull the test image and run it locally. Then you can use `composer run test:integration:dev` to invoke phpunit. This should work.

Thanks for the info! I'll give that a go.

File: src/components/NavigationAccountExpandCollapse.vue (hunk 2 above)
Comment (ChristophWurst) [reply src/components/NavigationAccountExpandCollapse.vue:61]: please use async and await (and a try-catch-finally) instead of explicit promises. You'll still find traces of promises in our code base but we generally try to use the modern coding conventions

File: src/store/getters.js [hunk 3]
```diff
@@ -31,6 +31,9 @@ export const getters = {
 	getAccount: (state) => (id) => {
 		return state.accounts[id]
 	},
+	getAccountSettings: (state) => (id = null) => {
+		return id ? state.accountSettings[id] : state.accountSettings
```
Comment (ChristophWurst) [inline src/store/getters.js:35]: I'm not sure if we also need the variant with id=null. Can we make the id mandatory and always return `state.accountSettings[id]`

File: src/store/mutations.js [hunk 4]
```diff
@@ -77,7 +77,7 @@ export default {
 		Vue.set(state.preferences, key, value)
 	},
 	addAccount(state, account) {
-		account.collapsed = true
+		account.collapsed = account.collapsed !== undefined ? account.collapsed : true
```
Comment (ChristophWurst) [inline src/store/mutations.js:80]: `account.collapsed ?? true` for short

File: src/store/mutations.js [hunk 5]
```diff
@@ -110,6 +110,11 @@ export default {
 	expandAccount(state, accountId) {
 		state.accounts[accountId].collapsed = false
 	},
+	updateAccountSetting(state, { accountId, key, value }) {
```
Comment (ChristophWurst) [inline src/store/mutations.js:113]: let's call it `setAccountSetting` because it also works with values that were not set before

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Thanks for this and congrats to your first ever FLOSS contribution :)

I've left a few comments and remarks. Generally I'm on board with your approach. I too think that we shouldn't pollute the accounts table with too much UI-specifics and preferences and I like the approach with the serialized object.

What we'll need to pay attention to is phantom updates. Like when one process updates the preferences, we shouldn't overwrite the full object but just patch the one property that is sent. This will prevent concurrency issues when we later have more than one account preference.

Please also see the ticket that I linked. You're implementing a feature that was requested in the past.

File: src/components/Navigation.vue (hunk 1 above)
Comment (brueckner) [reply src/components/Navigation.vue:46]: Ah, I forgot to mention that. Yea, this is unrelated indeed. I've been getting a duplicate key warning:

Mail_-_Nextcloud

Comment (ChristophWurst) [reply src/components/Navigation.vue:46]: Ah now I see the issue. We display accounts and mailboxes. Hence there are indeed conflicts because they can have an overlap of IDs.

Good catch. Mind submtiting this as separate PR? :)

Comment (brueckner) [reply src/components/Navigation.vue:46]: Not at all :) I'll take care of the other comments first and will submit this separately afterwards.

File: src/store/getters.js (hunk 3 above)
Comment (brueckner) [reply src/store/getters.js:35]: Hm, I am using it [here](https://github.com/nextcloud/mail/pull/3952/files#diff-e01eb159ce5ee0617f7ca141ba6abc5eaf0dd0e4e9c89bfd77979ff31f63c0b6R153) to grab the settings for all accounts to store them. Maybe a separate method like `getAllAccountSettings` for clarity's sake?

Comment (ChristophWurst) [reply src/store/getters.js:35]: Oh I see. I didn't consider that `savePreference` is *global*, as in that it's not specific for one account. Hence you have to send this in a batch.

But yes, let's have a getter for this specific object and one for each account's settings for clarity :+1:

Comment (brueckner) [reply src/store/getters.js]: Hm, so... I wasn't actually using `getAccountSettings()` for anything other than fetching ALL account settings after a refactor yesterday. I renamed the method to `getAllAccountSettings()` in order to be specific and removed the parameter. If a need for something like `getAccountSettings(account.id)` comes up, we could add it down the line. Right now it would be unused.

Comment (ChristophWurst) [reply src/store/getters.js]: Makese sense, thanks

**Comment by brueckner** [comment]
> > I wanted to pass an empty array as a default fallback to `getPreference()` in [`PageController.php`](https://github.com/nextcloud/mail/compare/master...dotplex:enhancement/store-collapsed-folder-settings?expand=1#diff-70e5309fcd3311d771b3db9c93490ea270fd5894769093765ec37edb68e5dd9b) like this: `$this->preferences->getPreference('account-settings', []);` – however... the return value would be `null` instead of the empty array. Is that intentional? Seems like a bug to me.
> 
> Indeed. This should work.

I am not sure if I have been seeing ghosts, but I just tried it again and it works as intended, so I [changed it](https://github.com/nextcloud/mail/pull/3952/commits/5bc84f5e08d58330940a1cb07d80ee11cc58059e) to use the proper fallback value instead of the ternary.

**Review by ChristophWurst — APPROVED** [review]
Code looks good but didn't test (yet) :)

**Review by st3iny — CHANGES_REQUESTED** [review]

File: src/store/actions.js [hunk 6]
```diff
@@ -148,6 +148,10 @@ export default {
 			return account
 		})
 	},
+	setAccountSetting({ commit, getters }, { accountId, key, value }) {
+		commit('setAccountSetting', { accountId, key, value })
+		savePreference('account-settings', JSON.stringify(getters.getAllAccountSettings))
```
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply src/store/actions.js:153]: We could use `Object.values()` indeed to strip the IDs

[the remark this answers is not in the document]
Comment (brueckner) [reply src/store/actions.js]: Thanks for the feedback! I just noticed that doing a rebase right away is not the best for incremental reviews... 🙄 

Anyway, I decided to save it as an array of objects now, instead of an object with stringified ids. So the stored array now looks like this:

```json
[{"accountId":75,"collapsed":true},{"accountId":76,"collapsed":false}]
```

When merging the settings with the account (in main.js), I am simply grabbing the proper ones using `find`:

```javascript
const settings = accountSettings.find(settings => settings.accountId === account.id)
```

Most notably, [`setAccountSetting`](https://github.com/nextcloud/mail/pull/3952/commits/a150c9867256abb1a17ed4021d29c8f3a5285dbd#diff-7c0816590b8c43968bdb997a4d7ab7a803f17f160a132376e0b3aa1c9859d2f5R113-R122) has changed quite a bit in the process:

```javascript
setAccountSetting(state, { accountId, key, value }) {
	const accountSettings = state.allAccountSettings.find(settings => settings.accountId === accountId)
	if (accountSettings) {
		accountSettings[key] = value
	} else {
		const newAccountSettings = { accountId }
		newAccountSettings[key] = value
		state.allAccountSettings.push(newAccountSettings)
	}
},
```

... which takes care of updating existing settings or pushing new settings if there haven't been any for an account yet.

Lastly, I changed `state.accountSettings` to `state.allAccountSettings` for consistency.

[the remark this answers is not in the document]
Comment (st3iny) [reply src/store/actions.js]: The settings are saved to the database and persist one reload. However, `state.allAccountSettings` is not set on load in `main.js` (only the account state via `addAccount()`).

You could do something like:

```diff
diff --git a/src/main.js b/src/main.js
index 409b7f6f..018906c2 100644
--- a/src/main.js
+++ b/src/main.js
@@ -76,6 +76,16 @@ const accountSettings = JSON.parse(Base64.decode(getPreferenceFromPage('account-
 const accounts = JSON.parse(Base64.decode(getPreferenceFromPage('serialized-accounts')))
 accounts.map(fixAccountId).forEach((account) => {
 	const settings = accountSettings.find(settings => settings.accountId === account.id)
+	if (settings) {
+		delete settings.accountId
+		Object.entries(settings).forEach(([key, value]) => {
+			store.commit('setAccountSetting', {
+				accountId: account.id,
+				key,
+				value,
+			})
+		})
+	}
 	store.commit('addAccount', { ...account, ...settings })
 })
 

```

[the remark this answers is not in the document]
Comment (brueckner) [reply src/store/actions.js]: Thanks Richard, I'll take that as-is. :-)

I made the change and pulled in the latest changes from master.

**Review by st3iny — APPROVED** [review]
Works fine
